File: SignalServiceKit/src/Util/DataSource.m
//
// Copyright 2017 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

#import "DataSource.h"
#import "MIMETypeUtil.h"
#import "NSData+Image.h"
#import "OWSError.h"
#import "OWSFileSystem.h"
#import <SignalCoreKit/NSString+OWS.h>
#import <SignalServiceKit/SignalServiceKit-Swift.h>

NS_ASSUME_NONNULL_BEGIN

#pragma mark -

@interface DataSourceValue ()

@property (nonatomic) NSData *data;
@property (nonatomic) NSString *fileExtension;
@property (atomic) BOOL isConsumed;

// These properties is lazily-populated.
@property (nonatomic, nullable) NSURL *cachedFileUrl;
@property (nonatomic, nullable) ImageMetadata *cachedImageMetadata;

@end

#pragma mark -

@implementation DataSourceValue

- (void)dealloc
{
    NSURL *_Nullable fileUrl = self.cachedFileUrl;
    if (fileUrl != nil) {
        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
            [OWSFileSystem deleteFileIfExists:fileUrl.path];
        });
    }
}

- (instancetype)initWithData:(NSData *)data
               fileExtension:(NSString *)fileExtension
{
    self = [super init];
    if (!self) {
        return self;
    }
    _data = data;
    _fileExtension = fileExtension;
    _isConsumed = NO;

    // Ensure that value is backed by file on disk.
    __weak DataSourceValue *weakValue = self;
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        [weakValue dataUrl];
    });

    return self;
}

+ (_Nullable id<DataSource>)dataSourceWithData:(NSData *)data
                                 fileExtension:(NSString *)fileExtension
{
    OWSAssertDebug(data);

    if (!data) {
        OWSFailDebug(@"data was unexpectedly nil");
        return nil;
    }

    return [[self alloc] initWithData:data fileExtension:fileExtension];
}

+ (_Nullable id<DataSource>)dataSourceWithData:(NSData *)data utiType:(NSString *)utiType
{
    NSString *fileExtension = [MIMETypeUtil fileExtensionForUTIType:utiType];
    return [[self alloc] initWithData:data fileExtension:fileExtension];
}

+ (_Nullable id<DataSource>)dataSourceWithOversizeText:(NSString *_Nullable)text
{
    if (!text) {
        return nil;
    }

    NSData *data = [text.filterStringForDisplay dataUsingEncoding:NSUTF8StringEncoding];
    return [[self alloc] initWithData:data fileExtension:kOversizeTextAttachmentFileExtension];
}

+ (id<DataSource>)emptyDataSource
{
    return [[self alloc] initWithData:[NSData new] fileExtension:@"bin"];
}

#pragma mark - DataSource

@synthesize sourceFilename = _sourceFilename;

- (void)setSourceFilename:(nullable NSString *)sourceFilename
{
    OWSAssertDebug(!self.isConsumed);
    _sourceFilename = sourceFilename.filterFilename;
}

- (nullable NSURL *)dataUrl
{
    OWSAssertDebug(self.data);
    OWSAssertDebug(!self.isConsumed);

    @synchronized(self)
    {
        if (!self.cachedFileUrl) {
            NSURL *fileUrl = [OWSFileSystem temporaryFileUrlWithFileExtension:self.fileExtension
                                                 isAvailableWhileDeviceLocked:YES];
            if ([self writeToUrl:fileUrl error:nil]) {
                self.cachedFileUrl = fileUrl;
            } else {
                OWSLogDebug(@"Could not write data to disk: %@", self.fileExtension);
                OWSFailDebug(@"Could not write data to disk.");
            }
        }

        return self.cachedFileUrl;
    }
}

- (NSUInteger)dataLength
{
    OWSAssertDebug(self.data);
    OWSAssertDebug(!self.isConsumed);
    return self.data.length;
}

- (BOOL)writeToUrl:(NSURL *)dstUrl error:(NSError **)outError
{
    OWSAssertDebug(self.data);
    OWSAssertDebug(!self.isConsumed);
    NSError *error = nil;
    if (![self.data writeToURL:dstUrl options:NSDataWritingAtomic error:&error]) {
        OWSFailDebug(@"Could not write data to disk: %@", error);
        if (outError != NULL) {
            *outError = error;
        }
        return NO;
    }
    return YES;
}

- (BOOL)moveToUrlAndConsume:(NSURL *)dstUrl error:(NSError **)outError
{
    OWSAssertDebug(!self.isConsumed);

    @synchronized(self) {
        OWSAssertDebug(!NSThread.isMainThread);
        // This method is meant to be fast. If _cachedFileUrl is nil,
        // we'll still lazily generate it and this method will work,
        // but it will be slower than expected.
        OWSAssertDebug(self->_cachedFileUrl != nil);

        NSURL *_Nullable srcUrl = self.dataUrl;
        if (srcUrl == nil) {
            if (outError != NULL) {
                *outError = OWSErrorMakeAssertionError(@"Missing data URL.");
            }
            return NO;
        }
        self->_cachedFileUrl = nil;
        self.isConsumed = YES;
        NSError *error = nil;
        if (![OWSFileSystem moveFileFrom:srcUrl to:dstUrl error:&error]) {
            OWSFailDebug(@"Could not write data with error: %@", error);
            if (outError != NULL) {
                *outError = error;
            }
            return NO;
        }
        return YES;
    }
}

- (BOOL)isValidImage
{
    OWSAssertDebug(!self.isConsumed);
    return [self.data ows_isValidImage];
}

- (BOOL)isValidVideo
{
    OWSAssertDebug(!self.isConsumed);
    if (![MIMETypeUtil isSupportedVideoFile:self.dataUrl.path]) {
        return NO;
    }
    OWSFailDebug(@"Are we calling this anywhere? It seems quite inefficient.");
    return [OWSMediaUtils isValidVideoWithPath:self.dataUrl.path];
}

- (nullable NSString *)mimeType
{
    OWSAssertDebug(!self.isConsumed);
    if (self.fileExtension == nil) {
        OWSFailDebug(@"failure: fileExtension was unexpectedly nil");
        return nil;
    }

    return [MIMETypeUtil mimeTypeForFileExtension:self.fileExtension];
}

- (BOOL)hasStickerLikeProperties
{
    OWSAssertDebug(!self.isConsumed);
    return [self.data ows_hasStickerLikeProperties];
}

- (ImageMetadata *)imageMetadata
{
    OWSAssertDebug(!self.isConsumed);

    @synchronized(self) {
        if (self.cachedImageMetadata != nil) {
            return self.cachedImageMetadata;
        }
        ImageMetadata *imageMetadata = [self.data imageMetadataWithPath:nil mimeType:self.mimeType ignoreFileSize:YES];
        self.cachedImageMetadata = imageMetadata;
        return imageMetadata;
    }
}

@end

#pragma mark -

@interface DataSourcePath ()

@property (nonatomic) NSURL *fileUrl;
@property (nonatomic, readonly) BOOL shouldDeleteOnDeallocation;
@property (atomic) BOOL isConsumed;

// These properties is lazily-populated.
@property (nonatomic) NSData *cachedData;
@property (nonatomic, nullable) ImageMetadata *cachedImageMetadata;

@end

#pragma mark -

@implementation DataSourcePath

- (void)dealloc
{
    if (self.shouldDeleteOnDeallocation && !self.isConsumed) {
        NSURL *fileUrl = self.fileUrl;
        if (!fileUrl) {
            OWSFailDebug(@"fileUrl was unexpectedly nil");
            return;
        }

        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
            NSError *error;
            BOOL success = [[NSFileManager defaultManager] removeItemAtURL:fileUrl error:&error];
            if (!success || error) {
                OWSCFailDebug(@"DataSourcePath could not delete file: %@, %@", fileUrl, error);
            }
        });
    }
}

- (nullable instancetype)initWithFileUrl:(NSURL *)fileUrl
              shouldDeleteOnDeallocation:(BOOL)shouldDeleteOnDeallocation
                                   error:(NSError **)error
{
    if (!fileUrl || ![fileUrl isFileURL]) {
        NSString *errorMsg = [NSString stringWithFormat:@"unexpected fileUrl: %@", fileUrl];
        *error = OWSErrorMakeAssertionError(errorMsg);
        return nil;
    }

    self = [super init];
    if (!self) {
        return self;
    }

    _fileUrl = fileUrl;
    _shouldDeleteOnDeallocation = shouldDeleteOnDeallocation;
    _isConsumed = NO;

    return self;
}

+ (_Nullable id<DataSource>)dataSourceWithURL:(NSURL *)fileUrl
                   shouldDeleteOnDeallocation:(BOOL)shouldDeleteOnDeallocation
                                        error:(NSError **)error
{
    return [[self alloc] initWithFileUrl:fileUrl
              shouldDeleteOnDeallocation:shouldDeleteOnDeallocation
                                   error:error];
}

+ (_Nullable id<DataSource>)dataSourceWithFilePath:(NSString *)filePath
                        shouldDeleteOnDeallocation:(BOOL)shouldDeleteOnDeallocation
                                             error:(NSError **)error
{
    OWSAssertDebug(filePath);

    if (!filePath) {
        NSString *errorMsg = [NSString stringWithFormat:@"unexpected filePath: %@", filePath];
        *error = OWSErrorMakeAssertionError(errorMsg);
        return nil;
    }

    NSURL *fileUrl = [NSURL fileURLWithPath:filePath];
    return [[self alloc] initWithFileUrl:fileUrl
              shouldDeleteOnDeallocation:shouldDeleteOnDeallocation
                                   error:error];
}

+ (_Nullable id<DataSource>)dataSourceWritingTempFileData:(NSData *)data
                                            fileExtension:(NSString *)fileExtension
                                                    error:(NSError **)error
{
    NSURL *fileUrl = [OWSFileSystem temporaryFileUrlWithFileExtension:fileExtension isAvailableWhileDeviceLocked:YES];
    [data writeToURL:fileUrl options:NSDataWritingFileProtectionCompleteUntilFirstUserAuthentication error:error];
    if (*error != nil) {
        return nil;
    }
    return [[self alloc] initWithFileUrl:fileUrl shouldDeleteOnDeallocation:YES error:error];
}

+ (_Nullable id<DataSource>)dataSourceWritingSyncMessageData:(NSData *)data error:(NSError **)error
{
    return [self dataSourceWritingTempFileData:data fileExtension:kSyncMessageFileExtension error:error];
}

#pragma mark - DataSource

@synthesize sourceFilename = _sourceFilename;

- (void)setSourceFilename:(nullable NSString *)sourceFilename
{
    OWSAssertDebug(!self.isConsumed);
    _sourceFilename = sourceFilename.filterFilename;
}

- (NSData *)data
{
    OWSAssertDebug(!self.isConsumed);
    OWSAssertDebug(self.fileUrl);

    @synchronized(self)
    {
        if (!self.cachedData) {
            self.cachedData = [NSData dataWithContentsOfFile:self.fileUrl.path];
        }
        if (!self.cachedData) {
            OWSLogDebug(@"Could not read data from disk: %@", self.fileUrl);
            OWSFailDebug(@"Could not read data from disk.");
            self.cachedData = [NSData new];
        }
        return self.cachedData;
    }
}

- (NSUInteger)dataLength
{
    OWSAssertDebug(!self.isConsumed);
    OWSAssertDebug(self.fileUrl);

    NSNumber *fileSizeValue;
    NSError *error;
    [self.fileUrl getResourceValue:&fileSizeValue
                            forKey:NSURLFileSizeKey
                             error:&error];
    if (error != nil) {
        OWSLogDebug(@"Could not read data length from disk: %@, %@", self.fileUrl, error);
        OWSFailDebug(@"Could not read data length from disk with error: %@", error);
        return 0;
    }

    return fileSizeValue.unsignedIntegerValue;
}

- (nullable NSURL *)dataUrl
{
    OWSAssertDebug(!self.isConsumed);
    return self.fileUrl;
}

- (BOOL)isValidImage
{
    OWSAssertDebug(!self.isConsumed);
    return [NSData ows_isValidImageAtUrl:self.fileUrl mimeType:self.mimeType];
}

- (BOOL)isValidVideo
{
    OWSAssertDebug(!self.isConsumed);
    if (self.mimeType != nil) {
        if (![MIMETypeUtil isSupportedVideoMIMEType:self.mimeType]) {
            return NO;
        }
    } else if (![MIMETypeUtil isSupportedVideoFile:self.dataUrl.path]) {
        return NO;
    }
    return [OWSMediaUtils isValidVideoWithPath:self.dataUrl.path];
}

- (BOOL)hasStickerLikeProperties
{
    OWSAssertDebug(!self.isConsumed);
    return [NSData ows_hasStickerLikePropertiesWithPath:self.dataUrl.path];
}

- (ImageMetadata *)imageMetadata
{
    OWSAssertDebug(!self.isConsumed);

    @synchronized(self) {
        if (self.cachedImageMetadata != nil) {
            return self.cachedImageMetadata;
        }
        ImageMetadata *imageMetadata = [NSData imageMetadataWithPath:self.dataUrl.path
                                                            mimeType:self.mimeType
                                                      ignoreFileSize:YES];
        self.cachedImageMetadata = imageMetadata;
        return imageMetadata;
    }
}

- (BOOL)writeToUrl:(NSURL *)dstUrl error:(NSError **)outError
{
    OWSAssertDebug(!self.isConsumed);
    OWSAssertDebug(self.fileUrl);
    NSError *error = nil;
    if (![NSFileManager.defaultManager copyItemAtURL:self.fileUrl toURL:dstUrl error:&error]) {
        OWSFailDebug(@"Could not write data with error: %@", error);
        if (outError != NULL) {
            *outError = error;
        }
        return NO;
    }
    return YES;
}

- (BOOL)moveToUrlAndConsume:(NSURL *)dstUrl error:(NSError **)outError
{
    OWSAssertDebug(!self.isConsumed);
    OWSAssertDebug(self.fileUrl);

    self.isConsumed = YES;

    NSError *error = nil;
    BOOL success = NO;
    if ([[NSFileManager defaultManager] isWritableFileAtPath:self.fileUrl.path]) {
        success = [OWSFileSystem moveFileFrom:self.fileUrl to:dstUrl error:&error];
    } else {
        OWSLogError(@"File was not writeable. Copying instead of moving.");
        success = [NSFileManager.defaultManager copyItemAtURL:self.fileUrl toURL:dstUrl error:&error];
    }
    if (!success) {
        if (outError != NULL) {
            *outError = error;
        }
        OWSFailDebug(@"Could not write data with error: %@", error);
    }
    return success;
}

- (nullable NSString *)mimeType
{
    OWSAssertDebug(!self.isConsumed);
    NSString *_Nullable fileExtension = self.fileUrl.pathExtension;
    return (fileExtension ? [MIMETypeUtil mimeTypeForFileExtension:fileExtension] : nil);
}

@end

NS_ASSUME_NONNULL_END

File: SignalServiceKit/Util/DataSource.swift
//
// Copyright 2024 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

import Foundation

/// A base class that abstracts away a source of NSData
/// and allows us to:
///
/// * Lazy-load if possible.
/// * Avoid duplicate reads & writes.
public protocol DataSource: AnyObject {

    var sourceFilename: String? { get set }

    //// Should not be called unless necessary as it can involve an expensive read.
    var data: Data { get }

    /// The URL for the data.  Should always be a File URL.
    ///
    /// Should not be called unless necessary as it can involve an expensive write.
    ///
    /// Will only return nil in the error case.
    var dataUrl: URL? { get }

    /// Will return zero in the error case.
    var dataLength: UInt { get }
    var isValidImage: Bool { get }
    var isValidVideo: Bool { get }
    var hasStickerLikeProperties: Bool { get }
    var imageMetadata: ImageMetadata? { get }

    func writeTo(_ dstUrl: URL) throws

    func consumeAndDelete() throws
}

// MARK: -

public class DataSourceValue: DataSource {
    public init(_ data: Data, fileExtension: String) {
        self.data = data
        self.fileExtension = fileExtension
        self._isConsumed = false

        // TODO: Figure out if it's actually necessary to write to disk here; objc was triggering background write to disk so we kept this behavior, but it may not be necessary.
        DispatchQueue.global().async { [weak self] in
            // relies on the side effect of this computed property writing to disk; don't shoot the messenger...
            _ = self?.dataUrl
        }
    }

    public convenience init?(_ data: Data, utiType: String) {
        guard let fileExtension = MimeTypeUtil.fileExtensionForUtiType(utiType) else {
            return nil
        }
        self.init(data, fileExtension: fileExtension)
    }

    public convenience init?(_ data: Data, mimeType: String) {
        guard let fileExtension = MimeTypeUtil.fileExtensionForMimeType(mimeType) else {
            return nil
        }
        self.init(data, fileExtension: fileExtension)
    }

    public convenience init(oversizeText: String) {
        let data = Data(oversizeText.filterForDisplay.utf8)
        self.init(data, fileExtension: MimeTypeUtil.oversizeTextAttachmentFileExtension)
    }

    deinit {
        if let _dataUrl {
            try? OWSFileSystem.deleteFileIfExists(url: _dataUrl)
        }
    }

    public let data: Data
    private let fileExtension: String

    private let lock = NSRecursiveLock()

    /// Should only be accessed while holding `lock`.
    private var _isConsumed: Bool
    private var isConsumed: Bool {
        get {
            return lock.withLock {
                return _isConsumed
            }
        }
        set {
            lock.withLock {
                _isConsumed = newValue
            }
        }
    }

    /// This property is lazily-populated.
    /// Should only be accessed while holding `lock`.
    private var _imageMetadata: ImageMetadata??
    public var imageMetadata: ImageMetadata? {
        return lock.withLock { () -> ImageMetadata? in
            owsAssertDebug(!_isConsumed)
            if let _imageMetadata {
                return _imageMetadata
            }
            let cachedImageMetadata = DataImageSource(data).imageMetadata(ignorePerTypeFileSizeLimits: true)
            _imageMetadata = cachedImageMetadata
            return cachedImageMetadata
        }
    }

    private var _sourceFilename: String?
    public var sourceFilename: String? {
        get {
            return _sourceFilename
        }
        set {
            owsAssertDebug(!isConsumed)
            _sourceFilename = newValue?.filterFilename()
        }
    }

    /// This property is lazily-populated.
    /// Should only be accessed while holding `lock`.
    private var _dataUrl: URL?
    public var dataUrl: URL? {
        lock.withLock {
            owsAssertDebug(!_isConsumed)
            if let _dataUrl {
                return _dataUrl
            }
            let fileUrl = OWSFileSystem.temporaryFileUrl(fileExtension: fileExtension, isAvailableWhileDeviceLocked: true)
            if (try? writeTo(fileUrl)) != nil {
                _dataUrl = fileUrl
            } else {
                owsFailDebug("Could not write data to disk.")
            }
            return _dataUrl
        }
    }

    public var dataLength: UInt {
        owsAssertDebug(!isConsumed)
        return UInt(data.count)
    }

    public func writeTo(_ dstUrl: URL) throws {
        owsAssertDebug(!isConsumed)

        do {
            try data.write(to: dstUrl, options: .atomic)
        } catch {
            owsFailDebug("Could not write data to disk: \(error)")
            throw error
        }
    }

    public func consumeAndDelete() throws {
        try lock.withLock {
            owsAssertDebug(!_isConsumed)
            _isConsumed = true
            guard let _dataUrl else { return }
            try OWSFileSystem.deleteFileIfExists(url: _dataUrl)
        }
    }

    public var isValidImage: Bool {
        owsAssertDebug(!isConsumed)
        return DataImageSource(data).ows_isValidImage
    }

    public var isValidVideo: Bool {
        owsAssertDebug(!isConsumed)
        guard let path = dataUrl?.path else {
            return false
        }
        owsFailDebug("Are we calling this anywhere? It seems quite inefficient.")
        do {
            try OWSMediaUtils.validateVideoExtension(ofPath: path)
            try OWSMediaUtils.validateVideoSize(atPath: path)
            try OWSMediaUtils.validateVideoAsset(atPath: path)
            return true
        } catch {
            return false
        }
    }

    public var hasStickerLikeProperties: Bool {
        owsAssertDebug(!isConsumed)
        return imageMetadata?.hasStickerLikeProperties ?? false
    }
}

// MARK: -

public class DataSourcePath: DataSource {
    public init(fileUrl: URL, shouldDeleteOnDeallocation: Bool) throws {
        guard fileUrl.isFileURL else {
            throw OWSAssertionError("unexpected fileUrl: \(fileUrl)")
        }

        self.fileUrl = fileUrl
        self.shouldDeleteOnDeallocation = shouldDeleteOnDeallocation
        self._isConsumed = false
    }

    public convenience init(filePath: String, shouldDeleteOnDeallocation: Bool) throws {
        let fileUrl = URL.init(fileURLWithPath: filePath)
        try self.init(fileUrl: fileUrl, shouldDeleteOnDeallocation: shouldDeleteOnDeallocation)
    }

    public convenience init(writingTempFileData: Data, fileExtension: String) throws {
        let fileUrl = OWSFileSystem.temporaryFileUrl(fileExtension: fileExtension, isAvailableWhileDeviceLocked: true)
        try writingTempFileData.write(to: fileUrl, options: .completeFileProtectionUntilFirstUserAuthentication)
        try self.init(fileUrl: fileUrl, shouldDeleteOnDeallocation: true)
    }

    public convenience init(writingSyncMessageData: Data) throws {
        try self.init(writingTempFileData: writingSyncMessageData, fileExtension: MimeTypeUtil.syncMessageFileExtension)
    }

    deinit {
        if shouldDeleteOnDeallocation && !_isConsumed {
            // In the ObjC code this would fire into a dispatch queue
            do {
                try FileManager.default.removeItem(at: fileUrl)
            } catch {
                owsFailDebug("DataSourcePath could not delete file: \(fileUrl), \(error)")
            }
        }
    }

    public let fileUrl: URL
    private let shouldDeleteOnDeallocation: Bool
    private let lock = NSRecursiveLock()

    private var _isConsumed: Bool
    private var isConsumed: Bool {
        get {
            return lock.withLock {
                return _isConsumed
            }
        }
        set {
            lock.withLock {
                _isConsumed = newValue
            }
        }
    }

    private var _sourceFilename: String?
    public var sourceFilename: String? {
        get {
            return _sourceFilename
        }
        set {
            owsAssertDebug(!isConsumed)
            _sourceFilename = newValue?.filterFilename()
        }
    }

    private var _data: Data?
    public var data: Data {
        lock.withLock {
            owsAssertDebug(!_isConsumed)
            if _data == nil {
                _data = NSData(contentsOfFile: fileUrl.path) as Data?
            }
            if _data == nil {
                owsFailDebug("Could not read data from disk.")
                _data = Data()
            }
            return _data!
        }
    }

    public var dataLength: UInt {
        owsAssertDebug(!isConsumed)
        do {
            let values = try fileUrl.resourceValues(forKeys: [.fileSizeKey])
            return UInt(values.fileSize!)
        } catch {
            owsFailDebug("Could not read data length from disk with error: \(error)")
            return 0
        }
    }

    public var dataUrl: URL? {
        owsAssertDebug(!isConsumed)
        return fileUrl
    }

    public var isValidImage: Bool {
        owsAssertDebug(!isConsumed)
        return (try? DataImageSource.forPath(fileUrl.path))?.ows_isValidImage ?? false
    }

    public var isValidVideo: Bool {
        owsAssertDebug(!isConsumed)
        do {
            try OWSMediaUtils.validateVideoExtension(ofPath: fileUrl.path)
            try OWSMediaUtils.validateVideoSize(atPath: fileUrl.path)
            try OWSMediaUtils.validateVideoAsset(atPath: fileUrl.path)
            return true
        } catch {
            return false
        }
    }

    public var hasStickerLikeProperties: Bool {
        owsAssertDebug(!isConsumed)
        return imageMetadata?.hasStickerLikeProperties ?? false
    }

    private var _imageMetadata: ImageMetadata??
    public var imageMetadata: ImageMetadata? {
        lock.withLock { () -> ImageMetadata? in
            owsAssertDebug(!_isConsumed)
            if let _imageMetadata {
                return _imageMetadata
            }
            let imageMetadata = (try? DataImageSource.forPath(fileUrl.path))?.imageMetadata(ignorePerTypeFileSizeLimits: true)
            _imageMetadata = imageMetadata
            return imageMetadata
        }
    }

    public func writeTo(_ dstUrl: URL) throws {
        owsAssertDebug(!isConsumed)
        do {
            try FileManager.default.copyItem(at: fileUrl, to: dstUrl)
        } catch {
            owsFailDebug("Could not write data with error: \(error)")
        }
    }

    public func consumeAndDelete() throws {
        try lock.withLock {
            owsAssertDebug(!_isConsumed)
            _isConsumed = true

            try OWSFileSystem.deleteFileIfExists(url: fileUrl)
        }
    }
}
